# Hyperparameter Finetuning

In [2]:
import numpy as np
import pandas as pd
import seaborn as sns

## A. The data
### A.1. Uploading from csv - after featuer selection process

In [3]:
df = pd.read_csv('BGP_FS_10062022.csv')

In [5]:
df

ID      lng  ChWt_Min_W3  ChWt_Min_W4  ChWt_Min_W5  ChWt_AVG_W1  \
0          1 -75.6972     0.000000     0.000000     0.000000     0.000000   
1          2 -75.6972     0.000000     0.000000     0.000000     0.000000   
2          3 -75.6960     0.000000     0.000000     0.000000     0.000000   
3          4 -75.6972     0.000000     0.000000     0.000000     0.446303   
4          5 -75.6960     0.000000     0.000000     0.000000     0.000000   
...      ...      ...          ...          ...          ...          ...   
10267  10268 -97.7431     7.502572     7.699669     4.919473     7.693074   
10268  10269 -97.7431     7.502572     7.699669     4.919473     7.693074   
10269  10270 -97.7431     7.502572     7.699669     4.919473     7.693074   
10270  10271 -97.7431     7.502572     7.699669     4.919473     7.693074   
10271  10272 -97.7431     7.502572     7.699669     4.919473     7.693074   

       ChWt_STD_W1  Min_seaLvlPressure_W2  Min_seaLvlPressure_W3  \
0         0.000000               6.912842               6.915525   
1         0.000000               6.912842               6.915525   
2         0.000000               6.912842               6.915525   
3         1.202956               6.912842               6.915525   
4         0.000000               6.912842               6.915525   
...            ...                    ...                    ...   
10267     5.555519               6.919387               6.920277   
10268     5.555519               6.919387               6.920277   
10269     5.555519               6.919387               6.920277   
10270     5.555519               6.919387               6.920277   
10271     5.555519               6.919387               6.920277   

       Min_seaLvlPressure_W4  ...  Avg_seaLvlPressure_W5  Stm_Max_W4.high  \
0                   6.905753  ...               6.919005                0   
1                   6.905753  ...               6.919005                0   
2                   6.905753  ...               6.919005                0   
3                   6.905753  ...               6.919005                0   
4                   6.905753  ...               6.919005                0   
...                      ...  ...                    ...              ...   
10267               6.918695  ...               6.927715                0   
10268               6.918695  ...               6.927715                0   
10269               6.918695  ...               6.927715                0   
10270               6.918695  ...               6.927715                0   
10271               6.918695  ...               6.927715                0   

       Stm_Max_W5.high  Stm_Min_W1.high  Stm_Min_W5.high  Stm_AVG_W1.high  \
0                    0                0                0                0   
1                    0                0                0                0   
2                    0                0                0                0   
3                    0                0                0                0   
4                    0                0                0                0   
...                ...              ...              ...              ...   
10267                0                0                0                0   
10268                0                0                0                0   
10269                0                0                0                0   
10270                0                0                0                0   
10271                0                0                0                0   

       Stm_AVG_W5.high  Stm_STD_W1.high  Stm_STD_W5.high  El_Nxt_Mnth_Avg_OC  
0                    0                0                0            4.057455  
1                    0                0                0            4.696928  
2                    0                0                0            3.069193  
3                    0                0                0            5.260250  
4                    0 

In [6]:
df.shape

(10272, 22)

In [9]:
df.dtypes

ID                         int64
lng                      float64
ChWt_Min_W3              float64
ChWt_Min_W4              float64
ChWt_Min_W5              float64
ChWt_AVG_W1              float64
ChWt_STD_W1              float64
Min_seaLvlPressure_W2    float64
Min_seaLvlPressure_W3    float64
Min_seaLvlPressure_W4    float64
Min_seaLvlPressure_W5    float64
Avg_seaLvlPressure_W4    float64
Avg_seaLvlPressure_W5    float64
Stm_Max_W4.high            int64
Stm_Max_W5.high            int64
Stm_Min_W1.high            int64
Stm_Min_W5.high            int64
Stm_AVG_W1.high            int64
Stm_AVG_W5.high            int64
Stm_STD_W1.high            int64
Stm_STD_W5.high            int64
El_Nxt_Mnth_Avg_OC       float64
dtype: object

In [73]:
#df_s = df.sample(n=1000, random_state=2242) df_s = df.sample(n=1000, random_state=2242) 

In [72]:
#df_s 100 rows for run time check

### A.2. Creating the split; train - test - X, y ( outcome)

In [10]:
from sklearn.metrics import mean_absolute_error
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.model_selection import train_test_split, GridSearchCV, RandomizedSearchCV
from sklearn.preprocessing import OneHotEncoder, LabelEncoder

In [45]:
vn=list(df.columns)
vn

['ID',
 'lng',
 'ChWt_Min_W3',
 'ChWt_Min_W4',
 'ChWt_Min_W5',
 'ChWt_AVG_W1',
 'ChWt_STD_W1',
 'Min_seaLvlPressure_W2',
 'Min_seaLvlPressure_W3',
 'Min_seaLvlPressure_W4',
 'Min_seaLvlPressure_W5',
 'Avg_seaLvlPressure_W4',
 'Avg_seaLvlPressure_W5',
 'Stm_Max_W4.high',
 'Stm_Max_W5.high',
 'Stm_Min_W1.high',
 'Stm_Min_W5.high',
 'Stm_AVG_W1.high',
 'Stm_AVG_W5.high',
 'Stm_STD_W1.high',
 'Stm_STD_W5.high',
 'El_Nxt_Mnth_Avg_OC']

In [46]:
vn.remove('El_Nxt_Mnth_Avg_OC') #seperate the outcome for y

In [47]:
vn

['ID',
 'lng',
 'ChWt_Min_W3',
 'ChWt_Min_W4',
 'ChWt_Min_W5',
 'ChWt_AVG_W1',
 'ChWt_STD_W1',
 'Min_seaLvlPressure_W2',
 'Min_seaLvlPressure_W3',
 'Min_seaLvlPressure_W4',
 'Min_seaLvlPressure_W5',
 'Avg_seaLvlPressure_W4',
 'Avg_seaLvlPressure_W5',
 'Stm_Max_W4.high',
 'Stm_Max_W5.high',
 'Stm_Min_W1.high',
 'Stm_Min_W5.high',
 'Stm_AVG_W1.high',
 'Stm_AVG_W5.high',
 'Stm_STD_W1.high',
 'Stm_STD_W5.high']

In [67]:
### split file into train and test partial data to check run time... 1000 rows
#X_train, X_test, y_train, y_test = train_test_split(df[vn], df['El_Nxt_Mnth_Avg_OC'], test_size=0.20, random_state=2242)


In [74]:
### split file into train and test
X_train, X_test, y_train, y_test = train_test_split(df[vn], df['El_Nxt_Mnth_Avg_OC'], test_size=0.20, random_state=2242)


In [75]:
X_train

ID         lng  ChWt_Min_W3  ChWt_Min_W4  ChWt_Min_W5  ChWt_AVG_W1  \
8035  8036  -75.697200     9.568215     0.000000     0.000000    11.403809   
3745  3746  -74.700000     0.000000     5.726546     0.000000    10.225884   
3242  3243  -75.697200     3.590056     4.092585     4.144148     5.219670   
6288  6289  -75.696000     8.091533     8.245634     0.000000    10.491127   
8012  8013  -75.697200     0.000000     0.000000     0.000000    11.156453   
...    ...         ...          ...          ...          ...          ...   
8368  8369  -93.255398     3.868543     4.199029     4.199029     6.886608   
3555  3556  -97.743100     8.129320     0.000000     0.000000     8.145361   
9271  9272  -81.379039     5.160140     4.721492     3.248097     6.985233   
2316  2317  -93.255398     0.000000     0.000000     0.000000     0.000000   
4308  4309 -111.928140     1.779113     2.015450     2.040558     4.915154   

      ChWt_STD_W1  Min_seaLvlPressure_W2  Min_seaLvlPressure_W3  \
8035    10.619849               6.913936               6.910751   
3745     8.801503               6.900429               6.902944   
3242     5.163713               6.909154               6.905853   
6288    10.045459               6.919585               6.914632   
8012    10.578147               6.913936               6.910751   
...           ...                    ...                    ...   
8368     6.743404               6.918398               6.905052   
3555     6.236755               6.913936               6.914830   
9271     6.197549               6.914334               6.916517   
2316     0.000000               6.920375               6.917012   
4308     4.825479               6.910950               6.913240   

      Min_seaLvlPressure_W4  ...  Avg_seaLvlPressure_W4  \
8035               6.920672  ...               6.929664   
3745               6.909254  ...               6.924891   
3242               6.895278  ...               6.921735   
6288               6.909154  ...               6.924599   
8012               6.920672  ...               6.929664   
...                     ...  ...                    ...   
8368               6.890202  ...               6.922457   
3555               6.920079  ...               6.930400   
9271               6.920178  ...               6.929389   
2316               6.905753  ...               6.924756   
4308               6.916913  ...               6.920151   

      Avg_seaLvlPressure_W5  Stm_Max_W4.high  Stm_Max_W5.high  \
8035               6.920118                0                0   
3745               6.919096                0                0   
3242               6.906317                1                1   
6288               6.926394                0                0   
8012               6.920118                0                0   
...                     ...              ...              ...   
8368               6.904126                1                1   
3555               6.921233                1                1   
9271               6.927385                0                0   
2316               6.919475                0                0   
4308               6.916605                0                0   

      Stm_Min_W1.high  Stm_Min_W5.high  Stm_AVG_W1.high  Stm_AVG_W5.high  \
8035                0                0                0                0   
3745                0                0                0                0   
3242                1                1                1                1   
6288                0                0                0                0   
8012                0                0                0                0   
...               ...              ...              ...              ...   
8368                0                1                0                1   
3555                0                0                1                1   
9271                0                0                0                0   
2316            

In [76]:
X_test

ID         lng  ChWt_Min_W3  ChWt_Min_W4  ChWt_Min_W5  ChWt_AVG_W1  \
716    717  -81.379039     4.721492     5.228750     5.696742     6.555614   
9718  9719  -81.379039     0.000000     0.000000     0.000000    11.674891   
1604  1605  -81.379039     0.000000     0.000000     0.000000     5.599194   
1407  1408  -93.255398     7.795235     7.420378     7.566389     8.501211   
321    322  -76.485292     0.000000     0.000000     0.000000    11.328157   
...    ...         ...          ...          ...          ...          ...   
2689  2690  -97.743100     7.315711     0.000000     7.240312     7.430310   
9070  9071 -111.928140     4.918543     5.463816     5.769793     6.687288   
3102  3103  -93.255398     0.000000     0.000000     0.000000     1.881923   
1321  1322 -111.928140     2.574199     4.067122     3.878884     4.790620   
5244  5245  -97.743100     7.380593     7.632412     0.000000     7.785939   

      ChWt_STD_W1  Min_seaLvlPressure_W2  Min_seaLvlPressure_W3  \
716      5.342894               6.921362               6.921264   
9718    11.306380               6.917409               6.923432   
1604     5.278652               6.922348               6.921855   
1407     7.095865               6.916418               6.915624   
321     11.340853               6.907455               6.923235   
...           ...                    ...                    ...   
2689     6.777910               6.918695               6.918003   
9070     6.070024               6.916814               6.914830   
3102     2.239464               6.918102               6.896897   
1321     4.311781               6.910551               6.912842   
5244     5.184778               6.917409               6.919387   

      Min_seaLvlPressure_W4  ...  Avg_seaLvlPressure_W4  \
716                6.919980  ...               6.923607   
9718               6.919881  ...               6.926275   
1604               6.922250  ...               6.925653   
1407               6.917409  ...               6.924671   
321                6.909653  ...               6.920807   
...                     ...  ...                    ...   
2689               6.910551  ...               6.928297   
9070               6.916517  ...               6.920852   
3102               6.904550  ...               6.924184   
1321               6.914830  ...               6.918887   
5244               6.921757  ...               6.925009   

      Avg_seaLvlPressure_W5  Stm_Max_W4.high  Stm_Max_W5.high  \
716                6.923547                0                0   
9718               6.932444                0                0   
1604               6.922129                0                0   
1407               6.929817                1                1   
321                6.928571                0                0   
...                     ...              ...              ...   
2689               6.925214                0                0   
9070               6.920437                0                0   
3102               6.923543                0                0   
1321               6.916916                0                0   
5244               6.922894                0                0   

      Stm_Min_W1.high  Stm_Min_W5.high  Stm_AVG_W1.high  Stm_AVG_W5.high  \
716                 0                0                0                0   
9718                0                0                0                0   
1604                0                0                0                0   
1407                1                1                1                1   
321                 0                0                0                0   
...               ...              ...              ...              ...   
2689                0                0                0                0   
9070                0                0                0                0   
3102                0                0                0                0   
1321            

In [77]:
y_train

8035    3.976860
3745    5.211103
3242    7.168494
6288    3.828060
8012    3.898116
          ...   
8368    5.926589
3555    6.905038
9271    5.549451
2316    3.329188
4308    4.228464
Name: El_Nxt_Mnth_Avg_OC, Length: 8217, dtype: float64

In [78]:
y_test

716     4.954986
9718    3.457125
1604    4.311459
1407    6.290821
321     5.201434
          ...   
2689    5.828954
9070    4.386922
3102    5.672221
1321    2.218136
5244    5.246784
Name: El_Nxt_Mnth_Avg_OC, Length: 2055, dtype: float64

## B. Random Search - randomly decide which parameters will have which value
### B.1.  Setting the random range and parameters

In [79]:
# Number of trees in random forest
n_estimators = [int(x) for x in np.linspace(start = 100, stop = 1000, num = 10)]
# Number of features to consider at every split
max_features = ['auto', 'sqrt']
# Maximum number of levels in tree
max_depth = [int(x) for x in np.linspace(10, 110, num = 11)]
max_depth.append(None)
# Minimum number of samples required to split a node
min_samples_split = [2, 3, 5, 10]
# Minimum number of samples required at each leaf node
min_samples_leaf = [2, 3, 4, 5]
# Method of selecting samples for training each tree
bootstrap = [True, False]

### B.2. Seting the greed and view of the random results

In [80]:
# Create the random grid
random_grid = {'n_estimators': n_estimators,
               'max_features': max_features,
               'max_depth': max_depth,
               'min_samples_split': min_samples_split,
               'min_samples_leaf': min_samples_leaf,
               'bootstrap': bootstrap}


print(random_grid)

{'n_estimators': [100, 200, 300, 400, 500, 600, 700, 800, 900, 1000], 'max_features': ['auto', 'sqrt'], 'max_depth': [10, 20, 30, 40, 50, 60, 70, 80, 90, 100, 110, None], 'min_samples_split': [2, 3, 5, 10], 'min_samples_leaf': [2, 3, 4, 5], 'bootstrap': [True, False]}


### B.3. Running the Rnadom forest model iterating with the greed selected (randomly) parameters

In [81]:
rf = RandomForestRegressor()

rf_random = RandomizedSearchCV(estimator = rf, param_distributions = random_grid, n_iter = 50, cv = 10, 
                               verbose=2, random_state=2242, n_jobs = -1)

# Fit the random search model
rf_random.fit(X_train, y_train)

Fitting 10 folds for each of 50 candidates, totalling 500 fits


RandomizedSearchCV(cv=10, estimator=RandomForestRegressor(), n_iter=50,
                   n_jobs=-1,
                   param_distributions={'bootstrap': [True, False],
                                        'max_depth': [10, 20, 30, 40, 50, 60,
                                                      70, 80, 90, 100, 110,
                                                      None],
                                        'max_features': ['auto', 'sqrt'],
                                        'min_samples_leaf': [2, 3, 4, 5],
                                        'min_samples_split': [2, 3, 5, 10],
                                        'n_estimators': [100, 200, 300, 400,
                                                         500, 600, 700, 800,
                                                         900, 1000]},
                   random_state=2242, verbose=2)

## C. Results
### C.1. Best parameters found by the model iteration:

In [82]:
rf_random.best_params_

{'n_estimators': 300,
 'min_samples_split': 3,
 'min_samples_leaf': 2,
 'max_features': 'auto',
 'max_depth': 20,
 'bootstrap': True}

### C.2. Evaluation function

In [83]:
def evaluate(model, test_features, test_labels):
    predictions = model.predict(test_features)
    errors = abs(predictions - test_labels)
    mae = 100 * np.mean(errors)
    print('Model Performance')
    print('Mean Absolute Error: {:0.4f}'.format(np.mean(errors)))
    return mae

### C.3. Run base model - befor the finetuning 


In [84]:
base_model = RandomForestRegressor(n_estimators = 1000, random_state = 4)
base_model.fit(X_train, y_train)
base_accuracy = evaluate(base_model, X_test, y_test)

Model Performance
Mean Absolute Error: 0.4782


# Base model Mean Absolut Error = 0.4782 !!!
### C.4. Compare fine-tunned model


In [85]:
best_random = rf_random.best_estimator_
random_accuracy = evaluate(best_random, X_test, y_test)

Model Performance
Mean Absolute Error: 0.4743


# Finetuning resutl - Mean Absolut Error = 0.4743 !!!

### C.5. Improvmet persentage after fine tuning:

In [86]:
print('Improvement of {:0.2f}%.'.format( 100 * (base_accuracy - random_accuracy) / base_accuracy))

Improvement of 0.83%.


# Improvment of 0.83% after applying finetuning !!